File: lists/find_intersection.cpp
#include <iostream>
#include <stdlib.h>
#include <math.h>

using namespace std;

// Quick node class
struct Node {
	int value;
	Node *next;
};
// Creates new nodes
struct Node* newNode(int data)
{
	struct Node* temp = new struct Node;
	temp->value = data;
	temp->next = NULL;
}

struct Node* find_intersection(Node *L1, Node *L2){
	if (L1 == NULL || L2 == NULL)
		return NULL;
	int size1=0;
	int size2=0;
	
	Node* tmp1 = L1;
	Node* tmp2 = L2;
	// find sizes of the lists
	while (tmp1->next != NULL){
		tmp1=tmp1->next;
		size1++;
	}
	while (tmp2->next != NULL){
		tmp2=tmp2->next;
		size2++;
	}
	tmp1=L1;
	tmp2=L2;
	int difference = abs(size1-size2);
	//make the lists the same size 
	if (size1 > size2){
		for (int x=0;x<difference;x++)
			tmp1=tmp1->next;
	}
	if (size2 > size1){
		for (int x=0;x<difference;x++)
			tmp2=tmp2->next;
	}
	// move both pointers through lists until they are the same node
	while (tmp1->next!=NULL && tmp2->next!= NULL){
		if (tmp1 == tmp2){
			cout << "success!\n";
			return tmp1;
		}
		tmp1=tmp1->next;
		tmp2=tmp2->next;
	}

	cout << "no intersection found" << endl;
	return NULL;
}

int main(){
	// create 2 Linked lists using Nodes 
	struct Node *head1 = new struct Node;
	struct Node *head2 = new struct Node;
	// Construct linked list 1->0->1->8
	head2 = newNode(1);
	head2->next = newNode(0); 
	head2->next->next = newNode(1);
	head2->next->next->next = newNode(8);

		// Construct linked list 1->2->1->8
		// where the second 1 is the intersection
	head1 = newNode(1);
	head1->next = newNode(2);
	head1->next->next = head2->next->next;

	find_intersection(head1,head2);

	return 0;
}

File: lists/python/find_intersection.py
from LinkedList import LinkedList
from LinkedList import build_ll_from_lst
from Node import Node

"""
Given pointers to the head nodes of 2 linked lists that merge 
together at some point, find the Node where the two lists merge.

Assumptions:
1) Head nodes will not be NULL
2) Lists will always merge at some point

Approaches:
1) Hashmap + 2 loops - Loop through first LL and add all object hash ids + value
to hashmap. Loop through second and check if ID is in hashmap. Return value associated
with first node found (since that would be first intersection)
2) O(1) space, but O(n^2) time? For each node in L1, loop through L2 and until NULL and check
for intersection. Keep going until first intersection is found. Average case not terrible...
3) Linear time? Recursive?

[List #1] a--->b--->c
                     \
                      x--->y--->z--->NULL
                     /
     [List #2] p--->q

 1
  \
   2--->3--->NULL
  /
 1
#Returns 2

1--->2
      \
       3--->Null
      /
     1
#Returns 3

"""

def find_intersection_hashmap(headA, headB):
	hashmap = {}
	while headA is not None:
		hashmap[id(headA)] = headA.value
		headA = headA.next

	while headB is not None:
		hash_id = id(headB)
		if hashmap.get(hash_id) is not None:
			return hashmap[hash_id]
		headB = headB.next
	return None



#Tests

def test_find_intersection_hashmap():
	fourth_node = Node("D")
	third_node = Node("C", fourth_node)
	second_node = Node("B", third_node)
	head1 = Node("A", second_node)

	second_node2 = Node("B2", third_node)
	head2 = Node("A2", second_node2)

	assert find_intersection_hashmap(head1, head2) == "C"



if __name__ == "__main__":
	test_find_intersection_hashmap()

